expand derivation examples (#9048)
NixOS/nix: doc/manual/src/language/derivations.md

[store derivation]: @docroot@/glossary.md#gloss-store-derivation

## Input attributes

### Required

- [`name`]{#attr-name} ([String](@docroot@/language/values.md#type-string))

  A symbolic name for the derivation.
  It is added to the [store path] of the corresponding [store derivation] as well as to its [output paths](@docroot@/glossary.md#gloss-output-path).

  [store path]: @docroot@/glossary.md#gloss-store-path

  > **Example**
  >
  > ```nix
  > derivation {
  >   name = "hello";
  >   # ...
  > }
  > ```
  >
  > The store derivation's path will be `/nix/store/<hash>-hello.drv`.

>   > The [output](#attr-outputs) paths will be of the form `/nix/store/<hash>-hello[-<output>]`

Reviewer: Do you want to mention how `out` is special?

Author: It is mentioned in the outputs section.

Reviewer: Oh is the `[...]` supposed to indicate that suffix may or may not appear?

  This allows the associated [store objects](@docroot@/glossary.md#gloss-store-object) and their [closures](@docroot@/glossary.md#gloss-closure) to be copied or garbage-collected separately.

  > **Example**
  >
  > Imagine a library package that provides a dynamic library, header files, and documentation.
  > A program that links against such a library doesn’t need the header files and documentation at runtime, and it doesn’t need the documentation at build time.
  > Thus, the library package could specify:
  >
  > ```nix
  > derivation {
  >   # ...
  >   outputs = [ "lib" "dev" "doc" ];
  >   # ...
  > }
  > ```
  >
  > This will cause Nix to pass environment variables `lib`, `dev`, and `doc` to the builder containing the intended store paths of each output.
  > The builder would typically do something like
  >
  > ```bash
  > ./configure \
  >   --libdir=$lib/lib \
  >   --includedir=$dev/include \
  >   --docdir=$doc/share/doc
  > ```
  >
  > for an Autoconf-style package.

  The symbolic name is appended to an output's store path, unless it is `out`.

  > **Example**
  >
  >
  > ```nix
  > derivation {
  >   name = "example";

>   >   outputs = [ "lib" "dev" "doc" ];

Reviewer: ```suggestion
  >   outputs = [ "lib" "dev" "doc" "out" ];
```

  >   # ...
  > }
  > ```
  >
  > The store derivation path will be `/nix/store/<hash>-example.drv`.
  > The output paths will be
  > - `/nix/store/<hash>-example-lib`
  > - `/nix/store/<hash>-example-dev`

>   > - `/nix/store/<hash>-example-doc`

Reviewer: ```suggestion
  > - `/nix/store/<hash>-example-doc`
  > - `/nix/store/<hash>-example`
```

  You can refer to each output of a derivation by selecting it as an attribute.
  The first element of `outputs` determines the *default output* and ends up at the top-level.

  > **Example**
  >
  > Select an output by attribute name:
  >
  > ```nix
  > let
  >   myPackage = derivation {
  >     name = "example";

>   >     outputs = [ "lib" "dev" "doc" ];

Author: ```suggestion
  >     outputs = [ "lib" "dev" "doc" "out" ];
```

  >     # ...
  >   };
  > in myPackage.dev
  > ```
  >

>   > SInce `lib` is the first output, `myPackage` is equivalent to `myPackage.lib`.

Author: ```suggestion
  > Since `lib` is the first output, `myPackage` is equivalent to `myPackage.lib`.
```

  By default, a derivation produces a single output called `out`.
  However, derivations can produce multiple outputs.
  This allows the associated [store objects](@docroot@/glossary.md#gloss-store-object) and their [closures](@docroot@/glossary.md#gloss-closure) to be copied or garbage-collected separately.

  > **Example**
  >
  > Imagine a library package that provides a dynamic library, header files, and documentation.
  > A program that links against such a library doesn’t need the header files and documentation at runtime, and it doesn’t need the documentation at build time.
  > Thus, the library package could specify:
  >
  > ```nix
  > derivation {
  >   # ...
  >   outputs = [ "lib" "dev" "doc" ];
  >   # ...
  > }
  > ```
  >
  > This will cause Nix to pass environment variables `lib`, `dev`, and `doc` to the builder containing the intended store paths of each output.
  > The builder would typically do something like
  >
  > ```bash
  > ./configure \
  >   --libdir=$lib/lib \
  >   --includedir=$dev/include \
  >   --docdir=$doc/share/doc
  > ```
  >
  > for an Autoconf-style package.

>   The symbolic name is appended to an output's store path, unless it is `out`.

Reviewer: ```suggestion
  The name of an output is combined with the name of the derivation to create the name part of the outputs store path, unless it is `out`, in which case just the name of the derivation is used.
```

Reviewer: I think it's good to be very explicit. "Symbolic name" I don't think is so clear, and the really is notion of output name just here but with e.g. `/nix/store/adsfasdfasdfasdf-drv^output` derived path syntax. There is no special case for `out` with that, for example.
